lib/prawn/document.rb

```diff
@@ -683,21 +683,20 @@ def generate_margin_box
     end
 
     def apply_margin_options(options)
-      if options[:margin]
-        # Treat :margin as CSS shorthand with 1-4 values.
-        margin = Array(options[:margin])
-        positions = { 4 => [0, 1, 2, 3], 3 => [0, 1, 2, 1],
-                      2 => [0, 1, 0, 1], 1 => [0, 0, 0, 0] }[margin.length]
-
-        [:top, :right, :bottom, :left].zip(positions).each do |p, i|
-          options[:"#{p}_margin"] ||= margin[i]
-        end
-      end
+      sides = [:top, :right, :bottom, :left]
+      margin = Array(options[:margin])
 
-      [:left, :right, :top, :bottom].each do |side|
-        if margin = options[:"#{side}_margin"]
-          state.page.margins[side] = margin
-        end
+      # Treat :margin as CSS shorthand with 1-4 values.
+      positions = { 4 => [0, 1, 2, 3], 3 => [0, 1, 2, 1],
+                    2 => [0, 1, 0, 1], 1 => [0, 0, 0, 0],
+                    0 => [] }[margin.length]
+
+      sides.zip(positions).each do |side, pos|
+        state.page.margins[side] = margin[pos] if pos
+
+        # side-specific margin options take precedence over :margin
+        specific = options["#{side}_margin"]
```

Comment:
Unless I'm misunderstanding you -- when there are no margins in the options hash, `pos` is nil each time through the loop. Trying to evaluate `margins[nil]` would raise a TypeError.

Two workarounds I noticed were supplying dummy values to the positions array, or looping over `sides.zip(Array(positions))` -- both seemed less explicit.